libz-rs-sys/src/gz.rs

```diff
@@ -1296,6 +1303,340 @@ unsafe fn gz_decomp(state: &mut GzState) -> Result<(), ()> {
+        unsafe { gz_error(state, Some((Z_MEM_ERROR, "out of memory"))) };
+        return Err(());
+    };
+    state.input = input.as_ptr();
+    // Note: zlib-ng fills the input buffer with zeroes here, but it's unneeded.
+
+    // Only need output buffer and deflate state if compressing.
+    if !state.direct {
+        // Allocate output buffer.
+        state.out_size = state.want;
+        let Some(output) = ALLOCATOR.allocate_slice_raw::<u8>(state.out_size) else {
+            unsafe { free_buffers(state) };
+            // Safety: The caller confirmed the validity of state.
+            unsafe { gz_error(state, Some((Z_MEM_ERROR, "out of memory"))) };
+            return Err(());
+        };
+        state.output = output.as_ptr();
+
+        // Allocate deflate memory, set up for gzip compression.
+        state.stream.zalloc = Some(ALLOCATOR.zalloc);
+        state.stream.zfree = Some(ALLOCATOR.zfree);
+        state.stream.opaque = ALLOCATOR.opaque;
+        const DEF_MEM_LEVEL: c_int = 8;
+        if unsafe {
+            deflateInit2_(
+                &mut state.stream,
+                state.level as _,
+                Z_DEFLATED,
+                MAX_WBITS + 16,
+                DEF_MEM_LEVEL,
+                state.strategy as _,
+                zlibVersion(),
+                core::mem::size_of::<z_stream>() as _,
+            )
+        } != Z_OK
+        {
+            unsafe { free_buffers(state) };
+            // Safety: The caller confirmed the validity of state.
+            unsafe { gz_error(state, Some((Z_MEM_ERROR, "out of memory"))) };
+            return Err(());
+        }
+        state.stream.next_in = ptr::null_mut();
+    }
+
+    // Note: zlib-ng sets state.size = state.want here to mark the state as initialized.
+    // We don't have state.size, so gz_write looks for a non-null state.input buffer
+    // (which we allocated above) to tell if the state has been initialized.
+
+    // Initialize write buffer if compressing.
+    if !state.direct {
+        state.stream.avail_out = state.out_size as _;
+        state.stream.next_out = state.output;
+        state.next = state.stream.next_out;
+    }
+
+    Ok(())
+}
+
+// Compress whatever is at avail_in and next_in (unless in direct mode) and write
+// to the output file.
+//
+// # Returns
+//
+// - `Ok` on success.
+// - `Err` on error.
+fn gz_comp(state: &mut GzState, flush: c_int) -> Result<(), ()> {
+    // Allocate memory if this is the first time through
+    if state.input.is_null() && gz_init(state).is_err() {
+        return Err(());
+    }
+
+    // Write directly if requested.
+    if state.direct {
+        let got = unsafe {
+            libc::write(
+                state.fd,
+                state.stream.next_in.cast::<c_void>(),
+                state.stream.avail_in as _,
+            )
+        };
+        if got < 0 || got as c_uint != state.stream.avail_in {
+            unsafe { gz_error(state, Some((Z_ERRNO, "write error"))) };
```

Comment:
The C version uses `zstrerror` here. You just assume that the error will be `"write error"` in this case?

Reply:
Yes, I left a FIXME comment about implementing `zstrerror` in the corresponding part of the read path, but I forgot to add it here.

Implementing `zstrerror` is probably big enough to live in a separate PR. zlib-ng aliases it to the C `strerror` function, which [isn't necessarily thread-safe](https://pubs.opengroup.org/onlinepubs/9699919799/functions/strerror.html). Probably we should use `strerror_r`, although that isn't available on all platforms (a scan of the Rust `libc` source suggests it's missing on Win32, for example). For now I'll add a FIXME comment here.